# Poisson Distribution Exploration

The Poisson distribution is used when we are interested in calculating the probability that an event will have a certain number of occurrences in a fixed interval of time. For example, the number of product defects that a factory will discover on any given day.

A Poisson distribution is characterized by a certain lambda value, $λ$. It so happens that under the Poisson distribution, this value is equal to both the __expected value__ of the distribution, as well as the __variance__ of the distribution. Formally, we define the Poisson distribution for a given $λ$ and a given number of occurrences $k$ as:

$$
P(X=k) = \frac{e^{-\lambda} \lambda^k}{k!}
$$

where $e$ is the base of the natural logarithm:
$$
e \approx 2.71828
$$

In the example given above, $λ$ is the expected number of defects to be discovered on any given day, although in reality, the number of actual defects discovered on a particular day could be different.

In this project, we will explore how we can use the properties of the Poisson distribution to calculate the probabilities of different numbers of defects being discovered. For this project, let's assume that $λ = 7$. Then the formula for our distribution becomes:

$$
P(X=k) = \frac{e^{-7} 7^k}{k!}
$$

We'll begin by importing the necessary Python libraries for the project: __numpy__ and the stats module within __scipy__. We'll also set $λ = 7$ to be used throughout the project. Using $λ$ instead of the number 7 will also help keep our code readable when we have several numerical arguments within a given method:

In [1]:
# Import numpy and scipy.stats
import numpy as np
import scipy.stats as stats

# Set λ = 7
λ = 7

Let's begin by calculating the probability that the number of defects found on a particular day will actually be equal to 7. Since the Poisson distribution is a _discrete_ probability distribution, it takes on a countable number of values. Therefore, we can use a probability mass function (PMF) to calculate the probability of an exact number of occurrences rather than only a range of values (as would be the case for a continuous random variable):

In [2]:
# Calculate the probability of λ under the Poisson distribution
prob_λ = stats.poisson.pmf(λ, λ) * 100

# Display prob_λ
prob_λ

14.900277967433773

Given that the expected value $\lambda$ of our distribution is 7, the probability that we actually encounter exactly 7 defects on any given day is 14.9%. Note that the above syntax includes multiplication by 100, so prob_λ is a _percentage_. Without the multiplication by 100, we would get a decimal value (.0149...) between 0 and 1, as would be expected.

An exceptionally good day for the factory occurs when no more than 4 defects are found in one day. What is the probability of this occurring?

We are now interested in the probability of 4 _or fewer_ defects, so we are interested in the probability of 0, 1, 2, 3, _or_ 4 defects. We could calculate the probability of each possibility individually using the probability mass function, but it will be more efficient to instead use a cumulative density function (CDF) with 4 as our given parameter. This will calculate the probability of all possible values for 4 or fewer:

$$
P(X \leq 4) = \sum_{i=0}^{4} \frac{e^{-7} \times 7^i}{i!}
$$

In [3]:
# Calculate the probability of 4 or less under the Poisson distribution
fewer_than_4 = stats.poisson.cdf(4, λ) * 100

# Display fewer_than_4
fewer_than_4

17.299160788207146

Again, we multiplied by 100, so the probability that the factory will find 4 or fewer defects in one day is 17.3%.

On the other hand, the factory views more than 9 defects as a bad day. What is the probability of this happening?

Note that the wording here tells us that 9 is _not_ to be included in our calculation. To most effectively calculate this, it is more advantageous to calculate the _complement_ of what we're asked. If the probability of event $A$ is given by $P(A)$, then the probability of the complement of $A$, denoted as $P(A')$ is given by:

$$ P(A') = 1 - P(A) $$


The complement of "more than 9" is "9 or fewer". We then subtract the probability of 9 or fewer defects being found from 1 to obtain the probability of more than 9 defects being found. 

We again use a cumulative density function to calculate the probability of 9 or fewer defects occurring and subtract this value from 1:

$$
P(X > 9) = 1 - \sum_{i=0}^{9} \frac{e^{-7} \times 7^i}{i!}
$$

In Python code, we write:

In [4]:
# Calculate the probability of 9 or more occurrences
more_than_9 = (1 - stats.poisson.cdf(9, λ)) * 100

# Display more_than_9
more_than_9

16.950406276132668

The probability of more than 9 occurrences on a given day is 16.95%.

Now that we've looked at a few examples of probabilities for only one day, let's look at probabilities on a larger scale. Let's say the factory wants to track information about the number of defects over 1 year. While a real factory would certainly not be open every day of the year, let's assume for simplicity's sake that it is. If the factory records the number of defects found at the end of the day, we'd have a list of 365 values. 

Let's create an experiment that simulates this.

First, we want to create a set of 365 randomly-generated values given the Poisson distribution with $λ = 7$. In this case, we'll use the random variate sample (RVS) method:

In [5]:
# Randomly generate 365 values under the given Poisson distribution
year_defects = stats.poisson.rvs(λ, size = 365)

# Display year_defects
year_defects

array([ 3,  9,  6,  5,  7,  8,  5,  6,  5,  7,  6,  5,  5,  6,  9,  4,  5,
        5,  6,  5, 10,  6, 12,  2, 11,  1,  5,  7,  9,  3, 11,  9,  7,  3,
        5,  4,  3,  7,  7,  2,  7,  7,  8,  5,  9, 10,  6,  6,  9,  2, 14,
        4,  7,  6,  7,  6,  5,  6,  9,  5,  3,  2,  4,  9,  5,  7,  6, 12,
        9,  5,  7,  8,  8,  2,  9,  5, 12,  3,  6,  7,  4,  7,  5,  5,  6,
        8,  5,  9,  7,  6,  7,  9,  7,  9, 10, 10,  8, 13, 11, 12,  9,  4,
       10,  6, 11, 11,  8,  7,  9,  9, 10,  7, 12,  3,  6, 12,  6,  6, 11,
       11,  9,  7,  4,  8,  2,  8,  7,  3,  9, 10,  4,  7, 12,  8,  9,  9,
        7,  6,  9,  8,  1,  6, 13, 12,  6,  9,  6, 10,  9,  3,  5,  2, 10,
        7,  9,  6, 12,  7,  9, 10,  7, 10,  4, 12,  8,  7,  4, 11, 12,  7,
        8,  8,  5,  7,  9, 10,  6,  9, 10,  9, 15,  9,  9,  8,  5,  5,  4,
        9,  7,  8, 11,  6,  6,  5, 10,  8,  4,  5,  3,  6,  6, 10,  4,  6,
        6, 11,  7,  6,  9, 11, 11,  5,  6,  8,  3,  6,  8, 10,  6,  4,  3,
        8, 11,  6,  5,  2

Because the values are randomly generated, running the function again will yield a different random set of 365 values:

In [6]:
# Randomly generate 365 values under the given Poisson distribution
year_defects = stats.poisson.rvs(λ, size = 365)

# Display year_defects
year_defects

array([ 8,  9,  5,  5,  6,  7,  5,  9,  4,  6,  6,  6,  9,  8, 13,  6,  5,
        7,  4,  8,  7,  9,  2, 11,  9,  8,  7,  8, 11,  3,  3,  3, 12,  8,
        4, 15, 13,  8, 11, 10,  4,  5,  5,  2,  6,  7,  9,  5,  7,  8, 11,
        4,  4,  2,  6,  5,  8, 12,  6, 10,  8,  8,  3,  9,  9,  7,  3,  8,
       10,  9,  4,  6,  8,  7,  8,  9, 11,  3,  4,  5,  6,  4,  7,  2,  5,
        7,  7,  9,  4, 12,  7,  8,  9,  7,  5, 10,  4,  7, 10, 13,  7,  8,
        9,  2,  5,  9,  3,  5,  8,  6,  9,  7,  9,  4, 10,  7,  9,  9,  7,
        8, 10,  4,  9,  8,  4,  5,  8, 11, 13,  4,  8, 13,  6,  7,  6,  7,
        8,  5,  7,  9,  4, 11,  6,  8,  6,  6,  6,  9,  7,  4,  8, 11, 11,
        7, 10,  6,  7, 10,  3,  7,  6,  6,  6,  5,  7,  9, 11,  7,  5, 12,
        6, 10,  5, 15,  6,  3,  5,  9,  9,  9,  9,  8,  6,  3, 12,  5,  6,
        8,  9,  5,  5,  7,  7,  6,  4,  8,  6,  7,  5,  7,  7,  8, 12, 11,
        8,  8,  3,  5,  4,  7, 10,  5,  8,  7, 10, 10,  2,  8, 12,  6, 11,
        8,  8,  5,  8,  5

Since we are simulating an experiment and want to make the results reproducible for a reader, we can choose a random seed value that will produce the same random variate sample each time we run it. We'll follow the custom of using a random seed value of 42 and redefine the variables year_defects after doing so:

In [7]:
# Set random seed = 42
np.random.seed(42)

# Randomly generate 365 values under the given Poisson distribution
year_defects = stats.poisson.rvs(λ, size = 365)

# Display year_defects
year_defects

array([ 6,  7,  6,  7,  7,  4, 12,  2,  7,  5,  7,  6, 10,  3,  6, 12,  5,
        6,  9,  6, 10,  4,  4,  7,  8, 11,  3,  2, 10,  6,  9,  6, 10, 10,
       14,  3,  5,  7,  8,  3,  7, 10,  2,  5,  6,  6,  7,  7,  3,  5, 12,
       10,  4, 12,  6,  4,  5, 11,  3,  3,  9,  6,  9,  9, 16,  5,  7, 13,
       10,  5,  3, 11,  8,  3,  4,  5,  3,  6,  6,  8,  7,  5,  7,  6,  8,
        4,  6, 10,  6,  6,  5, 10,  7,  7,  9,  9,  9,  8,  8,  7,  9,  8,
        8,  3,  8,  7,  4,  5,  5,  7,  7,  4,  4,  5,  9,  2,  8,  8,  6,
        8,  4,  4,  9,  4,  7,  6,  6,  8,  5, 12,  8, 10,  4,  6,  6, 16,
        4,  1,  8, 11,  5, 10,  6,  9,  6, 14,  8,  8,  8,  9,  6,  7,  9,
        8, 10,  6,  7,  8,  6,  1,  7,  5,  4,  3,  6,  4,  9,  8,  6, 11,
        6,  7, 11,  8,  8,  7,  8,  7,  4,  6,  5,  8,  7, 10, 10, 15,  3,
        7,  8, 10,  3,  5,  7,  6,  9, 11,  7,  6,  5, 10,  5,  5,  8,  9,
        5, 11,  8,  6, 10,  9, 11,  4,  5,  6,  5,  6,  4, 10,  5,  4,  3,
        9,  7,  9,  9,  4

Now, no matter how many times we run the code in cell 14, "year_defects" will always be assigned the same exact values. We demonstrate this below:

In [8]:
# Set random seed = 42
np.random.seed(42)

# Randomly generate 365 values under the given Poisson distribution
year_defects = stats.poisson.rvs(λ, size = 365)

# Display year_defects
year_defects

array([ 6,  7,  6,  7,  7,  4, 12,  2,  7,  5,  7,  6, 10,  3,  6, 12,  5,
        6,  9,  6, 10,  4,  4,  7,  8, 11,  3,  2, 10,  6,  9,  6, 10, 10,
       14,  3,  5,  7,  8,  3,  7, 10,  2,  5,  6,  6,  7,  7,  3,  5, 12,
       10,  4, 12,  6,  4,  5, 11,  3,  3,  9,  6,  9,  9, 16,  5,  7, 13,
       10,  5,  3, 11,  8,  3,  4,  5,  3,  6,  6,  8,  7,  5,  7,  6,  8,
        4,  6, 10,  6,  6,  5, 10,  7,  7,  9,  9,  9,  8,  8,  7,  9,  8,
        8,  3,  8,  7,  4,  5,  5,  7,  7,  4,  4,  5,  9,  2,  8,  8,  6,
        8,  4,  4,  9,  4,  7,  6,  6,  8,  5, 12,  8, 10,  4,  6,  6, 16,
        4,  1,  8, 11,  5, 10,  6,  9,  6, 14,  8,  8,  8,  9,  6,  7,  9,
        8, 10,  6,  7,  8,  6,  1,  7,  5,  4,  3,  6,  4,  9,  8,  6, 11,
        6,  7, 11,  8,  8,  7,  8,  7,  4,  6,  5,  8,  7, 10, 10, 15,  3,
        7,  8, 10,  3,  5,  7,  6,  9, 11,  7,  6,  5, 10,  5,  5,  8,  9,
        5, 11,  8,  6, 10,  9, 11,  4,  5,  6,  5,  6,  4, 10,  5,  4,  3,
        9,  7,  9,  9,  4

Let's explore these values a bit more. First, let's calculate the _expected_ number of defects to be found given a Poisson distribution with $λ = 7$. Since we're interested in how many defects we expect to find in 365 days, we just multiple $λ$ by 365:

In [9]:
# Calculate the expected number of defects in one year
λ * 365

2555

Under our given Poisson distribution, we would expect to find 2555 defects.

How does our randomly generated list compare?

In [10]:
# Calculate the total number of defects found in the list year_defects 
np.sum(year_defects)

2534

This does not seem extraordinarily far off from the expected value above, but since this is a _random_ sample, we should expect to see some variation. Let's see what happens when we change the value of our random seed to 10:

In [11]:
# Set the random seed value to 10
np.random.seed(10)

# Generate a new random sample under the Poisson distribution
year_defects2 = stats.poisson.rvs(λ, size = 365)

# Generate year_defects2
year_defects2

array([ 5,  6, 10, 10,  4,  9,  7,  7,  4,  6,  7,  6,  9,  7,  5,  6,  5,
        8, 10,  7,  4, 12,  7,  6,  7,  9,  7,  2,  6,  9,  9,  8,  9,  8,
        4,  8,  7,  6,  6,  6,  4,  5,  8,  9,  6,  6,  7,  8,  7,  3,  2,
        6, 11,  7,  6,  9,  7,  8,  1,  8, 13,  7,  4,  4,  3,  5,  4, 10,
        5,  4,  5,  4,  2, 10,  2,  4,  6,  9,  6,  5,  7,  8,  8,  4,  4,
        9,  5,  4,  6, 11,  8, 12,  3,  8, 10,  7,  7,  7,  9,  4,  9, 11,
        8,  7, 13,  9,  8,  7,  4,  6,  9,  7,  6, 10,  6,  8,  5, 12,  6,
       10, 10,  4,  7,  4,  6,  4,  7,  3,  6,  7,  6,  8,  6,  7,  7,  5,
        5,  9,  8,  9,  4,  9,  5,  8,  8,  9,  6,  2,  1, 10,  5,  7,  6,
        9, 10,  6, 10,  3,  4, 11,  7,  8,  3,  8, 13,  7,  7,  6,  6,  8,
        2,  9,  7,  4,  8, 10,  5, 13,  3,  7,  7,  5,  6, 10,  7, 11,  9,
        9,  8,  8,  8,  8, 10,  6,  5,  6,  3, 11,  6,  7,  7,  4, 14, 11,
       11,  6,  7,  4,  6,  9,  4,  6,  8,  9,  5,  5,  5,  8,  4,  7,  6,
        5,  8,  5, 13,  3

Now let's calculate the new sum:

In [12]:
# Calculate the sum of our new set
np.sum(year_defects2)

2462

This is a difference of almost 100 - much more substantial. Let's also calculate the mean of both lists:

In [13]:
# Calculate the means
np.mean(year_defects), np.mean(year_defects2)

(6.942465753424657, 6.745205479452054)

Neither are quite equal to our $λ$ value of 7. Should we be worried about this?

Since we've generated lists one at a time, we should expect to see some variation. However, by the Law of Large Numbers, if we generate enough lists, we should expect the calculated values to converge to the theoretical values. Let's explore this by generating 500 different samples:

In [14]:
# Create an empty list
year_lists = []

# Set the seed value to 42
np.random.seed(42)

# Create a "for" loop that generates 500 sample sets under our distribution and append each set to year_lists
for i in range(500):
    rvs = stats.poisson.rvs(λ, size = 365)
    year_lists.append(rvs)

Now let's find the total number of defects for each of the 500 trials:

In [15]:
# Create an empty list to store our sums
list_of_sums = []

# Create a "for" loop to loop over the elements of year_lists, finding the sum of each list and appending it to list_of_sums
for list in year_lists:
    sum = np.sum(list)
    list_of_sums.append(sum)

# Calculate the average sum
np.mean(list_of_sums)

2555.976

We see that the average number of defects across our 500 simulations is almost exactly our expected value.

Let's return to our 1st random generated list, year_defects, with with a random seed value of 42:

In [16]:
# Set random seed = 42
np.random.seed(42)

# Randomly generate 365 values under the given Poisson distribution
year_defects = stats.poisson.rvs(λ, size = 365)

# Display year_defects
year_defects

array([ 6,  7,  6,  7,  7,  4, 12,  2,  7,  5,  7,  6, 10,  3,  6, 12,  5,
        6,  9,  6, 10,  4,  4,  7,  8, 11,  3,  2, 10,  6,  9,  6, 10, 10,
       14,  3,  5,  7,  8,  3,  7, 10,  2,  5,  6,  6,  7,  7,  3,  5, 12,
       10,  4, 12,  6,  4,  5, 11,  3,  3,  9,  6,  9,  9, 16,  5,  7, 13,
       10,  5,  3, 11,  8,  3,  4,  5,  3,  6,  6,  8,  7,  5,  7,  6,  8,
        4,  6, 10,  6,  6,  5, 10,  7,  7,  9,  9,  9,  8,  8,  7,  9,  8,
        8,  3,  8,  7,  4,  5,  5,  7,  7,  4,  4,  5,  9,  2,  8,  8,  6,
        8,  4,  4,  9,  4,  7,  6,  6,  8,  5, 12,  8, 10,  4,  6,  6, 16,
        4,  1,  8, 11,  5, 10,  6,  9,  6, 14,  8,  8,  8,  9,  6,  7,  9,
        8, 10,  6,  7,  8,  6,  1,  7,  5,  4,  3,  6,  4,  9,  8,  6, 11,
        6,  7, 11,  8,  8,  7,  8,  7,  4,  6,  5,  8,  7, 10, 10, 15,  3,
        7,  8, 10,  3,  5,  7,  6,  9, 11,  7,  6,  5, 10,  5,  5,  8,  9,
        5, 11,  8,  6, 10,  9, 11,  4,  5,  6,  5,  6,  4, 10,  5,  4,  3,
        9,  7,  9,  9,  4

Let's say we want to know on what value $x$ the 90% percentile falls: in other words, on 90% of days, we can expect to find $x$ or fewer defects. This is identical to asking the question: for what value $x$ will the cumulative distribution function of $x$ under a given probability distribution be equal to .90: 

$$P(X \leq x) = 0.90 = \sum_{k=0}^{x} \frac{e^{-7} 7^k}{k!}$$



In Python, we're asking for what value $x$ will the running the code stats.poisson.cdf(x, λ) output .90?

In scipy.stats, we can use the probability point function (PPF) to do this:

In [17]:
# Set x equal to the probability point function of .90 under the given Poisson distribution
x = stats.poisson.ppf(.90, λ)

# Display x
x

10.0

We verify this result by entering 10 as our parameter into scipy's CDF method:

In [18]:
# Calculate the cumulative distribution function at 10 of our Poisson distribution
stats.poisson.cdf(10, λ)

0.9014792058890873

This is the same as saying that the 90th percentile of the data is 10, or that 90% of the time we should encounter 10 or fewer defects on any given day.

Let's look at the maximum and minimum of this data set and calculate the probability of getting each one:

In [19]:
# Calculate the maximum and minimum of our list year_defects and their respective CDFs
max = np.max(year_defects)
min = np.min(year_defects)
cdf_max = stats.poisson.cdf(max, λ)
cdf_min = stats.poisson.cdf(min, λ)

# Print the results
print(f"The probability of finding more than {max} defects or fewer is {(1 - cdf_max) *100:.2f}%.")
print(f"The probability of finding more than {min} defect or fewer is {(1 - cdf_min) * 100:.2f}%.")

The probability of finding more than 16 defects or fewer is 0.10%.
The probability of finding more than 1 defect or fewer is 99.27%.


In this project, we explore a number of properties of the Poisson distribution&mdash;a discrete probability distribution that can be used to calculate the probability of a particular number of occurrences of an event in a certain period of time given a mean value. In particular, we've see we can use the random variate sample method in scipy.stats to simulate real-life data and compare it to theoretical values and how these values become identical under the Law of Large Numbers. We've also seen how the probability point function is the inverse of the cumulative density function, allowing us to calculate percentiles under the distribution.